### ISSUE TITLE
Stackscript for Linode strangeness

### ISSUE THREAD

**Issue by chasd00** [progrium/dokku#1226]
1. The "Xenify distro" option wasn't in my profile config. I don't know if Linode has renamed it or what. I opened a ticket with them. When I set "Automount devtmpfs" to no I was able to boot.

2. The stackscript had a couple of errors like this:

E: Some index files failed to download. They have been ignored, or old ones used instead.
E: dpkg was interrupted, you must manually run 'sudo dpkg --configure -a' to correct the problem.
=> 06/10 21:36:12 Installing pre-requisites

When I manually did the dpkg --config -a then re-ran the stackscript it got further but never completed.

**Comment by chasd00** [comment]
1. Got  a response from Linode "Xenify distro" has been replaced with "Distro helper"

**Comment by michaelshobbs** [comment]
@josegonzalez thoughts here?

**Comment by chasd00** [comment]
Per this blogpost the Linode kernels have what they need to run Dokku out of the box.
https://blog.linode.com/2014/01/03/docker-on-linode/

Here's what I did to get Dokku running at Linode on Ubuntu 14.04 LTS after the free KVM upgrade
1. login as root and do $ssh-keygen
2. curl --silent https://get.docker.io/gpg 2> /dev/null | apt-key add - 2>&1 >/dev/null
3. curl --silent https://packagecloud.io/gpg.key 2> /dev/null | apt-key add - 2>&1 >/dev/null
4. echo "deb http://get.docker.io/ubuntu docker main" > /etc/apt/sources.list.d/docker.list
5. echo "deb https://packagecloud.io/dokku/dokku/ubuntu/ trusty main" > /etc/apt/sources.list.d/dokku.list
6. apt-get update > /dev/null
7. apt-get install -qq -y dokku 
- Ignore the modprobe error
8. reboot and finish through the web install (I chose the web install option)

I didn't have to mess with any of the profile options either ( "Xenify distro" or "Distro Helper" )

**Comment by nicksergeant** [comment]
Confirmed that @chasd00's steps worked for me too, though I switched to Linode's KVM implementation instead of Xen (that's what they're pushing new servers on eventually anyway).

**Comment by mrluc** [comment]
@chasd00's comment seems like a fairly important note for people using Linode! Seems like the documentation should be changed to reflect that a stackscript isn't required for those on KVM, which is what Linode calls "the future here at Linode."

(I followed the instructions in the docs about using the Stackscript, and it didn't end up "just working" at any rate -- I chose to use the web installer, and dokku never showed it to me. The install appears to have succeeded, for some definition of success -- I can push to it, but it makes things available at port numbers, instead of subdomains. I dug around in the docs quite a bit to see how to rectify this, and so far I have got *nothing*. But I'll open a new issue for that.)

**Comment by josegonzalez** [comment]
I haven't updated the stackscript in quite a while - it costs me money to screw around with Linode's setup - so if you want to dig into that, that would be quite awesome.

Would also be great to see these notes as documentation pull requests...

**Comment by mrluc** [comment]
Yeah - I wanted to actually test some things out for myself first. I did, and:

- @chasd00's approach worked because he manually ran the bootstrap's steps, _without_ trying to install `linux-image-extras`, which doesn't work no linode's ubuntu14.04 except through pv-grub+stackscript as mentioned.
- The bootstrap could run as-is on Linode if it didn't essentially require AUFS through linux-image-extras (with, of course, a note that operations will take from 5x to 10x as long -- starting/stopping containers, applying layers, etc).
- I just added a pull request that would make linux-image-extras optional, just in case we want discussion about that.
- The original "weirdness" in the stackscript appears to be solely due to the outdated naming of the radio-toggle it tells us to set in Linode's web interface; with that option set, the stackscript works as normal.

So this could prolly be closed with just a doc tweak.

**Comment by nicksergeant** [comment]
FWIW I just tried again using the StackScript w/ updated docs for the Distro Helper. I get through to the first email asking me to make those changes and reboot, but then nothing happens. My StackScript log:

```
root@dokkun:~# cat stackscript.log
=> 09/12 23:31:30 Setting root ssh key
=> 09/12 23:31:30 Turn off password authentication and root login for SSH
ssh stop/waiting
ssh start/running, process 3102
=> 09/12 23:31:30 Installing and configuring Postfix
=> 09/12 23:31:38 Set up hostname
=> 09/12 23:31:38 Installing docker gpg key
=> 09/12 23:31:39 Installing dokku gpg key
=> 09/12 23:31:39 Running apt-get update
=> 09/12 23:31:47 Installing pre-requisites
=> 09/12 23:31:48 Setting up apt repositories
=> 09/12 23:31:48 Running apt-get update
=> 09/12 23:31:56 Installing via linode
Preconfiguring packages ...
Selecting previously unselected package libfreetype6:amd64.
(Reading database ... 23605 files and directories currently installed.)
Preparing to unpack .../libfreetype6_2.5.2-1ubuntu2.5_amd64.deb ...
Unpacking libfreetype6:amd64 (2.5.2-1ubuntu2.5) ...
Selecting previously unselected package linux-image-3.13.0-63-generic.
Preparing to unpack .../linux-image-3.13.0-63-generic_3.13.0-63.103_amd64.deb ...
Done.
Unpacking linux-image-3.13.0-63-generic (3.13.0-63.103) ...
Selecting previously unselected package grub-common.
Preparing to unpack .../grub-common_2.02~beta2-9ubuntu1.3_amd64.deb ...
Unpacking grub-common (2.02~beta2-9ubuntu1.3) ...
Selecting previously unselected package grub2-common.
Preparing to unpack .../grub2-common_2.02~beta2-9ubuntu1.3_amd64.deb ...
Unpacking grub2-common (2.02~beta2-9ubuntu1.3) ...
Selecting previously unselected package grub-pc-bin.
Preparing to unpack .../grub-pc-bin_2.02~beta2-9ubuntu1.3_amd64.deb ...
Unpacking grub-pc-bin (2.02~beta2-9ubuntu1.3) ...
Selecting previously unselected package grub-pc.
Preparing to unpack .../grub-pc_2.02~beta2-9ubuntu1.3_amd64.deb ...
Unpacking grub-pc (2.02~beta2-9ubuntu1.3) ...
Selecting previously unselected package grub-gfxpayload-lists.
Preparing to unpack .../grub-gfxpayload-lists_0.6_amd64.deb ...
Unpacking grub-gfxpayload-lists (0.6) ...
Selecting previously unselected package linux-headers-3.13.0-63.
Preparing to unpack .../linux-headers-3.13.0-63_3.13.0-63.103_all.deb ...
Unpacking linux-headers-3.13.0-63 (3.13.0-63.103) ...
Selecting previously unselected package linux-headers-3.13.0-63-generic.
Preparing to unpack .../linux-headers-3.13.0-63-generic_3.13.0-63.103_amd64.deb ...
Unpacking linux-headers-3.13.0-63-generic (3.13.0-63.103) ...
Selecting previously unselected package linux-headers-generic.
Preparing to unpack .../linux-headers-generic_3.13.0.63.71_amd64.deb ...
Unpacking linux-headers-generic (3.13.0.63.71) ...
Selecting previously unselected package linux-headers-virtual.
Preparing to unpack .../linux-headers-virtual_3.13.0.63.71_amd64.deb ...
Unpacking linux-headers-virtual (3.13.0.63.71) ...
Selecting previously unselected package linux-image-virtual.
Preparing to unpack .../linux-image-virtual_3.13.0.63.71_amd64.deb ...
Unpacking linux-image-virtual (3.13.0.63.71) ...
Selecting previously unselected package linux-virtual.
Preparing to unpack .../linux-virtual_3.13.0.63.71_amd64.deb ...
Unpacking linux-virtual (3.13.0.63.71) ...
Selecting previously unselected package os-prober.
Preparing to unpack .../os-prober_1.63ubuntu1_amd64.deb ...
Unpacking os-prober (1.63ubuntu1) ...
Processing triggers for ureadahead (0.100.0-16) ...
Processing triggers for man-db (2.6.7.1-1ubuntu1) ...
Processing triggers for install-info (5.2.0.dfsg.1-2) ...
Setting up libfreetype6:amd64 (2.5.2-1ubuntu2.5) ...
Setting up linux-image-3.13.0-63-generic (3.13.0-63.103) ...
Running depmod.
update-initramfs: deferring update (hook will be called later)
Examining /etc/kernel/postinst.d.
run-parts: executing /etc/kernel/postinst.d/apt-auto-removal 3.13.0-63-generic /boot/vmlinuz-3.13.0-63-generic
run-parts: executing /etc/kernel/postinst.d/initramfs-tools 3.13.0-63-generic /boot/vmlinuz-3.13.0-63-generic
update-initramfs: Generating /boot/initrd.img-3.13.0-63-generic
run-parts: executing /etc/kernel/postinst.d/update-notifier 3.13.0-63-generic /boot/vmlinuz-3.13.0-63-generic
Setting up grub-common (2.02~beta2-9ubuntu1.3) ...
Setting up linux-headers-3.13.0-63 (3.13.0-63.103) ...
Setting up linux-headers-3.13.0-63-generic (3.13.0-63.103) ...
Setting up linux-headers-generic (3.13.0.63.71) ...
Setting up linux-headers-virtual (3.13.0.63.71) ...
Setting up linux-image-virtual (3.13.0.63.71) ...
Setting up linux-virtual (3.13.0.63.71) ...
Setting up os-prober (1.63ubuntu1) ...
Processing triggers for ureadahead (0.100.0-16) ...
Setting up grub2-common (2.02~beta2-9ubuntu1.3) ...
Setting up grub-pc-bin (2.02~beta2-9ubuntu1.3) ...
Setting up grub-gfxpayload-lists (0.6) ...
Setting up grub-pc (2.02~beta2-9ubuntu1.3) ...

Creating config file /etc/default/grub with new version
Processing triggers for libc-bin (2.19-0ubuntu6.3) ...
(Reading database ... 49805 files and directories currently installed.)
Removing grub-gfxpayload-lists (0.6) ...
Removing grub-pc (2.02~beta2-9ubuntu1.3) ...
Purging configuration files for grub-pc (2.02~beta2-9ubuntu1.3) ...
Processing triggers for man-db (2.6.7.1-1ubuntu1) ...
Preconfiguring packages ...
(Reading database ... 49783 files and directories currently installed.)
Removing grub2-common (2.02~beta2-9ubuntu1.3) ...
Processing triggers for man-db (2.6.7.1-1ubuntu1) ...
Processing triggers for install-info (5.2.0.dfsg.1-2) ...
Selecting previously unselected package grub.
(Reading database ... 49765 files and directories currently installed.)
Preparing to unpack .../grub_0.97-29ubuntu66_amd64.deb ...
Unpacking grub (0.97-29ubuntu66) ...
Processing triggers for man-db (2.6.7.1-1ubuntu1) ...
Setting up grub (0.97-29ubuntu66) ...
Searching for GRUB installation directory ... found: /boot/grub
Searching for default file ... Generating /boot/grub/default file and setting the default boot entry to 0
Searching for GRUB installation directory ... found: /boot/grub
Testing for an existing GRUB menu.lst file ...

Could not find /boot/grub/menu.lst file.
Generating /boot/grub/menu.lst
Searching for splash image ... none found, skipping ...
Found kernel: /boot/vmlinuz-3.13.0-63-generic
Updating /boot/grub/menu.lst ... done

Searching for GRUB installation directory ... found: /boot/grub
Searching for default file ... found: /boot/grub/default
Testing for an existing GRUB menu.lst file ... found: /boot/grub/menu.lst
Searching for splash image ... none found, skipping ...
Found kernel: /boot/vmlinuz-3.13.0-63-generic
Replacing config file /run/grub/menu.lst with new version
Updating /boot/grub/menu.lst ... done

=> 09/12 23:32:19 Sending notification email to nick@nicksergeant.com of required restart
```

**Comment by chasd00** [comment]
I'll try to work on the stackscript since I have a Linode account. The steps I originally posted are out of date I think, here's a copy/paste from a  document i put together to remind myself how to install dokku-alt from scratch on Linode, I think the steps ought to be similar for regular dokku.

Initial Setup
Choose a datacenter nearest to the user base
Once the server is finished being setup click the link to edit the configuration profile
Choose 64bit Ubuntu Linux 14.04 and leave the other settings set to their default.
Choose a root password, wisely
Set “Distro Helper” to off

After the server has been built, click Boot
Use the IP address assigned and set up DNS to properly point a subdomain to this host
For example, if apps are to be deployed at app1.test.mydomain.com
A record for test.mydomain.com -> IP address
CNAME alias *.test -> test.domain.com.

Install and configure Dokku-Alt
In Linode click on remote access and use the Lish ssh link to login to the new server as root
Edit /etc/hostname and /etc/hosts to properly reflect the fully qualified domain name for this server.
Run ssh-keygen, accept all the defaults
Run bash -c "$(curl -fsSL https://raw.githubusercontent.com/dokku-alt/dokku-alt/master/bootstrap.sh)"
Accept all defaults ( select first device in Grub ASCII menu )
In a browser open http://<server ip>:2000
Copy/paste the appropriate public key
Check use “virtual host naming”
Click “Finish Setup”

Run Ctrl-C to kill the setup webserver running in the console
Setup the firewall
Run ufw allow ssh/tcp
Run ufw allow http
Run ufw allow https
Run ufw enable

At the point the server is setup and ready.

**Comment by callahad** [comment]
Digging into this today.

The central issue seems to be that:

1. By default, Linode runs its own kernel which does *not* support AUFS.
2. The magic Docker shell script still really wants AUFS.
3. Therefore, if we want to use AUFS, we need to install Grub and boot a custom kernel.
4. (Bonus: Linode is in the middle of a transition from Xen to KVM, so we don't even know what hypervisor we'll run under.)

This is kind of tricky since we have to do initial provisioning, tell the user to manually tweak their Linode configuration, and then resume installing Docker and Dokku. StackScripts can't configure the kernel and aren't really designed to persist across a reboot boundary, so we have to use ugly hacks to run a self-deleting initscript on the next boot. And we still have to tell the user to go muck settings in between, which means this isn't really meaningfully automated.

I see a few options here, in order of preference:

1. __Deprecate the StackScript and update the docs.__ We walk people through booting their Linode onto the distro's kernel, at which point our own bootstrap.sh can take over.
2. __Use DeviceMapper instead of AUFS.__ This would let us run on Linode's standard kernels, which means our bootstrap.sh would work fine. Or we could write a simplified StackScript.
3. __Maintain the facade.__ As far as I can tell, we're currently failing to resume installation after reboot, so this will take some time and effort. And users still have to manually intervene between boots to swap out the kernel setting in the Linode dashboard.

Given that supporting Linode costs the core devs money and StackScripts seem to be inadequate for what we want to do, I'm most in favor of options that let us drop the StackScript altogether and funnel people through our normal, maintained channels.

Furthermore, Docker wants to move to the mainline kernel's overlay filesystem in a year or so, which would make this whole song and dance moot: given a modern kernel, our bootstrap.sh would just work.

**Comment by josegonzalez** [comment]
@callahad good call. Mind making a pr for #1 (including explaining linode's aufs/devicemapper situation)?

**Comment by josegonzalez** [comment]
Note: I'd be more than happy to fix the stackscript to be a bit more unattended but honestly I don't get paid for any of this so I am hesitant to spend my own money on a hosting provider I no longer use 

> Linode was good to me for a long time, and I miss their Longview. Thankfully I can more or less replicate that with cacti/graphite/grafana/diamond...

**Comment by callahad** [comment]
FWIW, the current bootstrap.sh script *does* work, it just dumps a scary warning when it starts up:

```
root@li1238-191:~# DOKKU_TAG=v0.4.1 bash bootstrap.sh
Preparing to install v0.4.1 from https://github.com/progrium/dokku.git...
--> Initial apt-get update
(Reading database ... 23405 files and directories currently installed.)
Preparing to unpack .../apt-transport-https_1.0.1ubuntu2.10_amd64.deb ...
Unpacking apt-transport-https (1.0.1ubuntu2.10) over (1.0.1ubuntu2.5) ...
Setting up apt-transport-https (1.0.1ubuntu2.10) ...
--> Installing docker
modprobe: ERROR: ../libkmod/libkmod.c:556 kmod_search_moddep() could not open moddep file '/lib/modules/4.1.5-x86_64-linode61/modules.dep.bin'
Warning: current kernel is not supported by the linux-image-extra-virtual
 package.  We have no AUFS support.  Consider installing the packages
 linux-image-virtual kernel and linux-image-extra-virtual for AUFS support.
+ sleep 10
[...continues per normal...]
```

Afterwards, Docker defaults to DeviceMapper and everything appears to work.

We could detect if the user is on a Linode default kernel (`uname -r | grep -q linode`) and drop a link to docs explaining the situation, sleep 10, then continue.

**Comment by josegonzalez** [comment]
I'm okay with that.
